docs/async_ops.md

```diff
@@ -0,0 +1,106 @@
+# Async HLO Instructions
+
+1. Adding async operations to HLO is cumbersome (i.e. `all-reduce-start` and
+   `all-reduce-done`).
+2. The start and done split may be inadequate for some of the asynchronous use
+   cases.
+
+To target the first shortcoming, we propose to introduce one last set of new
+asynchronous opcodes: `kAsyncStart`, `kAsyncUpdate`, and `kAsyncDone``. The idea
+is to create a generic asynchronous opcode that can wrap any HLO instruction.
+The actual operation that will be performed asynchronously will be encoded using
+a called computation that only has the instruction as its root and any
+parameters for inputs. The in-flight input/output buffer handling and aliasing
+can then be shared for any asynchronous operation. The async-start instruction’s
+output shape will then be a tuple of the input operands, output values, and any
+intermediate state that is needed for the `async-update` or `async-done`
+instructions.
+
+```
+%async_op {
+  %param0 = f32[64] parameter(0)
+  ROOT %op = f32[32] op(f32[64] %param0), op_specific_attr=”foo”
+}
+
+%async-start = (f32[64], f32[32], s32[]) async-start(f32[64] %operand),
+                                         calls=%async_op
+%async-done = f32[32] async-done((f32[64], f32[32], s32[]) %async-start), calls=%async_op
+```
+
+In the representation above, only `async-start` has a called computation since
+it is trivial to find what the `async-done` does by following its operand to
```

Comment:
@ezhulenev The example async-done as a calls attribute. This contradict the text here.